### PR TITLE
[^459] refactor(hx-checkbox): forms epic refactor hx-checkox and simplify

### PR DESCRIPTION
**Pull request by cathysiller**
JIRA: SURF-1453

### LGTM's
- [x] Dev LGTM
- [x] Design LGTM
- [x] Zoom LGTM

### CODE REVIEW

**Comment by HelixUI** [comment]
Deploy preview for *helix-ui* processing.

Building with commit fdf556019110a260a10305e9469f16d2e1c4d475

https://app.netlify.com/sites/helix-ui/deploys/5c62ee2c9b5e8b000b1cf89d

File: docs/components/checkboxes/checkbox-demo.js [hunk 1]
```diff
@@ -29,3 +29,9 @@ if (document.getElementById('vue-checkboxDemo')) {
         },
     });
 }
+
+let elIndeterminate = document.getElementById('chkIndeterminate');
```
Comment (CITguy) [inline docs/components/checkboxes/checkbox-demo.js]: Please revert this change.

File: docs/docs.js [hunk 2]
```diff
@@ -6,7 +6,6 @@ import './components/accordions/accordion-demo';
 import './components/alerts/alert-demo';
 import './components/box/box-demo';
 import './components/buttons/button-demo';
-import './components/checkboxes/checkbox-demo';
```
Comment (CITguy) [inline docs/docs.js]: Please revert this change.

File: src/helix-ui/elements/HXCheckboxElement.less [hunk 3]
```diff
@@ -24,20 +25,22 @@
...
-  }
+#hxMinus {
+  display: var(--hxMinus-display);
+}
+
+#hxTick {
+  display: var(--hxTick-display);
 }
 
 #hxNativeControl:checked:not(:indeterminate) ~ #hxCustomControl #hxTick {
```
Comment (CITguy) [inline src/helix-ui/elements/HXCheckboxElement.less]: You can probably remove everything from line 46 onward. This is old CSS that no longer applies to the new Shadow DOM markup.

File: src/helix-ui/styles/elements/hx-checkbox.less [hunk 4]
```diff
@@ -1,10 +1,132 @@
+@import 'mixins';
```
Comment (CITguy) [inline src/helix-ui/styles/elements/hx-checkbox.less]: This file should be moved to `components/checkbox.less`, at some point.

File: src/helix-ui/styles/elements/hx-div.less [hunk 5]
```diff
@@ -79,9 +79,9 @@ hx-div {
   }
 }
 
-// ===== Modern Browsers =====
+// =====  Browsers =====
 // Should only apply to browsers with ShadowDOM support.
-@supports(--modern: true) {
+@supports(--: true) {
```
Comment (CITguy) [inline src/helix-ui/styles/elements/hx-div.less]: Please revert changes to this file. This looks like an accidental change.

File: src/helix-ui/styles/components/checkbox.less [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+    margin: 0;
+    padding: 0;
+    //user-select: none;
+
+    > hx-checkbox {
+      -ms-grid-column: 1;
+      -ms-grid-row: 1;
+      grid-area: facade;
+      display: none;
+      margin-right: 0.50rem;
```
Comment (CITguy) [inline src/helix-ui/styles/components/checkbox.less]: this should be on the native `input`

File: src/helix-ui/styles/components/checkbox.less [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+    }
+
+    > span {
```
Comment (CITguy) [inline src/helix-ui/styles/components/checkbox.less]: do we still need this? double check IE

File: src/helix-ui/styles/components/checkbox.less [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+      grid-area: txt;
+      line-height: 1.5;
+    }
+  }
+
+  // help text
+  > p {
+    -ms-grid-column: 2;
+    -ms-grid-row: 2;
+    font-size: 0.75rem;
```
Comment (CITguy) [inline src/helix-ui/styles/components/checkbox.less]: add `color: @gray-750;` to adhere to spec

File: src/helix-ui/styles/components/checkbox.less [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+    grid-area: help;
+    margin: 0;
+    order: 2;
+  }
+
+  > input[type="checkbox"] {
+    background-color: transparent;
```
Comment (CITguy) [inline src/helix-ui/styles/components/checkbox.less]: does this have any effect?

File: src/helix-ui/styles/components/checkbox.less [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+    &:focus {
+      box-shadow: @focus-glow;
+    }
+
+    &:hover {
+      color: @cyan-500;
+
+      &::-ms-check {
+        background-color: @cyan-50;
+        border-color: @cyan-500;
```
Comment (CITguy) [inline src/helix-ui/styles/components/checkbox.less]: would `color: @cyan-500` work, instead?

File: src/helix-ui/styles/components/checkbox.less [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+        &::-ms-check {
+          border-color: @red-900;
+          color: @red-900;
+        }
+      }
+
+      &:hover {
+        &::-ms-check {
+          background-color: @red-100;
+          border-color: @red-500;
```
Comment (CITguy) [inline src/helix-ui/styles/components/checkbox.less]: change `border-color` to reflect design specs

File: src/helix-ui/styles/components/checkbox.less [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+        }
+      }
+    }
+
+    &:disabled {
+      color: @gray-500;
+      cursor: not-allowed;
+
+      &::-ms-check {
+        background-color: @gray-100 !important;
```
Comment (CITguy) [inline src/helix-ui/styles/components/checkbox.less]: please add comment to why we're using `!important`

File: src/helix-ui/styles/components/checkbox.less [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+          border-color: @gray-500;
+          color: @gray-500;
+        }
+      }
+    }
+  }
+}
+
+
+@supports (display: grid) {
```
Comment (CITguy) [inline src/helix-ui/styles/components/checkbox.less]: double check if block still needed

File: src/helix-ui/styles/components/checkbox.less [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+  hx-checkbox-control {
+    > label > hx-checkbox {
+      margin-right: 0;
+    }
+  }
+}
+
+@supports (display: inline-grid) {
```
Comment (CITguy) [inline src/helix-ui/styles/components/checkbox.less]: `s/inline-grid/grid/`

File: src/helix-ui/styles/components/checkbox.less [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+  hx-checkbox-control {
+    > input[type="checkbox"] {
+      #Mix.hideInput();
+      border: 0;
+      grid-area: ctrl;
```
Comment (CITguy) [inline src/helix-ui/styles/components/checkbox.less]: remove `grid-area`, already set further up the file

File: src/helix-ui/styles/components/checkbox.less [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+
+      &:checked {
+        ~ label > hx-checkbox {
+          --hxCheckbox__tick-display: block;
```
Comment (CITguy) [inline src/helix-ui/styles/components/checkbox.less]: add line for the minus display, to ensure it's not visible in this state

File: src/helix-ui/styles/components/checkbox.less [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+    > label > hx-checkbox {
+      margin-right: 0;
+    }
+  }
+}
+
+@supports (display: inline-grid) {
+  hx-checkbox-control {
+    > input[type="checkbox"] {
+      #Mix.hideInput();
```
Comment (CITguy) [inline src/helix-ui/styles/components/checkbox.less]: Let's refactor this to keep as much of the state styling together (outside of the `@supports`). Styles in the `@supports` should be very minimal.

File: src/helix-ui/styles/components/checkbox.less (hunk 7 above)
Comment (cathysiller) [reply src/helix-ui/styles/components/checkbox.less]: double checked ie, not needed/removed

File: src/helix-ui/styles/components/checkbox.less (hunk 9 above)
Comment (cathysiller) [reply src/helix-ui/styles/components/checkbox.less]: no effect, removed

File: src/helix-ui/styles/components/checkbox.less (hunk 10 above)
Comment (cathysiller) [reply src/helix-ui/styles/components/checkbox.less]: yes that works, updated

File: src/helix-ui/styles/components/checkbox.less (hunk 13 above)
Comment (cathysiller) [reply src/helix-ui/styles/components/checkbox.less]: doesnt look like its needed, removed

File: src/helix-ui/styles/components/checkbox.less (hunk 16 above)
Comment (cathysiller) [reply src/helix-ui/styles/components/checkbox.less]: updated

File: src/helix-ui/styles/components/checkbox.less [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+  }
+}
+
+/* ========== INVALID ========== */
+hx-checkbox-control > input[type="checkbox"]:invalid {
+  color: @red-900;
+
+  &::-ms-check {
+    border-width: 2px;
+  }
```
Comment (100stacks) [inline src/helix-ui/styles/components/checkbox.less]: would it be useful to put the `border-color` and `color` with `!important`?

File: src/helix-ui/styles/components/checkbox.less [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+
+    &:hover {
+      background-color: @red-100;
+      border-color: @red-900;
+      color: @red-900;
+    }
+  }
+
+  &:focus {
+    ~ label > hx-checkbox {
```
Comment (100stacks) [inline src/helix-ui/styles/components/checkbox.less]: combine with the `focus` block above.

File: src/helix-ui/styles/components/checkbox.less [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+}
+
+/* ========== DISABLED ========== */
+hx-checkbox-control > input[type="checkbox"]:disabled {
+  color: @gray-500;
+  cursor: not-allowed;
+
+  &::-ms-check {
+    background-color: @gray-100 !important; // simplest way of styling
+    border-width: 1px;
```
Comment (100stacks) [inline src/helix-ui/styles/components/checkbox.less]: be aware of invalid importance.

File: docs/components/checkboxes/index.html [hunk 21]
```diff
@@ -12,38 +13,66 @@ <h2 id="basic-checkbox">Basic Checkbox</h2>
...
+              </label>
+            </hx-checkbox-control>
+            <hx-checkbox-control>
+              <input id="chkIndeterminate" type="checkbox" v-model="isIndeterminate" />
+              <label for="chkIndeterminate">
+                <hx-checkbox></hx-checkbox>
+                Indeterminate
+              </label>
+              {% raw %}
+              <p class="hxSubBody">
```
[the remark this answers is not in the document]
Comment (cathysiller) [reply docs/components/checkboxes/index.html]: not needed, removed

File: src/helix-ui/elements/HXCheckboxElement.less [hunk 22]
```diff
@@ -1,128 +1,33 @@
...
+#hxCheckbox {
   align-items: center;
-  background-color: @gray-0;
-  border: 1px solid currentColor;
-  border-radius: 2px;
-  color: @gray-500;
-  display: flex;
-  font-size: 0.625rem; /* ~10px */
+  display: grid;
+  font-size: 10px; /* ~10px */
```
[the remark this answers is not in the document]
Comment (cathysiller) [reply src/helix-ui/elements/HXCheckboxElement.less]: removed

File: src/helix-ui/styles/components/checkbox.less [hunk 23]
```diff
@@ -0,0 +1,250 @@
+@import (reference) 'mixins';
+@import (reference) 'vars';
+
+// ========== BASE ELEMENTS ==========
+hx-checkbox {
+  background-color: @gray-0;
```
[the remark this answers is not in the document]
Comment (cathysiller) [reply src/helix-ui/styles/components/checkbox.less]: moved

File: src/helix-ui/styles/components/checkbox.less [hunk 24]
```diff
@@ -0,0 +1,250 @@
...
+  border-radius: 2px;
+  border: 1px solid @gray-500;
+  display: inline-block;
+  height: 1rem;
+  vertical-align: baseline;
```
[the remark this answers is not in the document]
Comment (cathysiller) [reply src/helix-ui/styles/components/checkbox.less]: not needed, removing

File: src/helix-ui/styles/components/checkbox.less [hunk 25]
```diff
@@ -0,0 +1,250 @@
...
+  }
+
+  > input[type="checkbox"] {
+    -ms-grid-column: 1;
+    -ms-grid-row: 1;
+    display: inline-block;
+    grid-area: ctrl;
+    height: 1rem;
+    margin-right: 0.50rem;
+    margin: 0;
```
[the remark this answers is not in the document]
Comment (cathysiller) [reply src/helix-ui/styles/components/checkbox.less]: abstracted

File: src/helix-ui/styles/components/checkbox.less [hunk 26]
```diff
@@ -0,0 +1,250 @@
...
+      color: @cyan-900;
+
+      ~ label > hx-checkbox {
+        border-color: @cyan-900;
+        color: @cyan-900;
+      }
+    }
+
+    &:valid {
+      // hx-checkbox-control > input[type="checkbox"]:valid >
```
[the remark this answers is not in the document]
Comment (cathysiller) [reply src/helix-ui/styles/components/checkbox.less]: removed, will add comments for FF

File: src/helix-ui/styles/components/checkbox.less (hunk 18 above)
Comment (cathysiller) [reply src/helix-ui/styles/components/checkbox.less]: yes! updated

File: src/helix-ui/styles/components/checkbox.less (hunk 19 above)
Comment (cathysiller) [reply src/helix-ui/styles/components/checkbox.less]: combined

File: src/helix-ui/styles/components/checkbox.less (hunk 20 above)
Comment (cathysiller) [reply src/helix-ui/styles/components/checkbox.less]: did this need updating?

File: docs/components/checkboxes/index.html [hunk 27]
```diff
@@ -3,6 +3,7 @@
 minver: 0.1.8
 also:
   components/checkboxes/test.html: Testing - Checkboxes
+  elements/hx-checkbox-control: <hx-checkbox-control>
```
Comment (CITguy) [inline docs/components/checkboxes/index.html]: Please remove link. It's currently broken.

**Comment by CITguy** [comment]
Also noticed a small CSS issue with Edge. The border of the `input::-ms-check` is still visible, so we'll need to turn off its border in the `@supports` block.

File: src/helix-ui/styles/components/checkbox.less [hunk 28]
```diff
@@ -0,0 +1,249 @@
...
+
+    > p {
+      margin-left: 1.5rem;
+    }
+
+    > input[type="checkbox"] {
+      #Mix.hideInput();
+    }
+
+    ::-ms-check {
```
Comment (CITguy) [inline src/helix-ui/styles/components/checkbox.less]: qualify this on the `input` ...

```less
> input[type="checkbox"] {
  &::-ms-check {
    border-width: 0;
  }
}
```

**Comment by araiford** [comment]
There's a significant vertical offset between the checkbox and the label text.  Ryan worked with me to step through options for getting that aligned, and it looks like there's a solution for now -- until we have a more solid vertical alignment strategy for inline items.

Once the solution is in place, this one should be good to go.

**Comment by araiford** [comment]
Checked out the adjustment.  Still about a half-pixel of offset, but this is closer and i've given it the LGTM.  Looking forward to how we handle vertical alignment in general.

Thank you @cathysiller and @CITguy !
